### DISCUSSION TITLE
Searching Model/Mixin Class Defining Specific Column in Declarative Table

### DISCUSSION THREAD

**Discussion by pykenny** [sqlalchemy/sqlalchemy#8609] [Usage Questions] [answered]
What's the best way to look up the model class / mixin that introduces given a column attribute in a declarative table?

Example:
```python
@declarative_mixin
class TimestampMixin(object):
    created_at = Column(DateTime, default=func.now())

class AttributeSetA(DeferredReflection, Base)
    __abstract__ = True
    column_a_1 = Column(Integer)
    column_a_2 = Column(String)

class AttributeSetB(AbstractConcreteBase, Base):
    column_b_1 = Column(Integer)
    column_b_2 = Column(String)

class MyModel(AttributeSetA, AttributeSetB, TimestampMixin):
    __tablename__ = "my_model"
    id = Column(Integer, primary_key=True)
```

From above, we know that column `id` is introduced by `MyModel`, `created_at` is introduced by `TimestampMixin`, and vice versa.

For `MyModel`, it's relatively simple to call `MyModel.__table__.c` to obtain all the actual columns corresponding to the model, but for abstract models / mixins without linkage to managed DB table (or more general, a mapper), this option is not available.

One of the possible (but maybe not so robust) ways to pull out columns from an abstract model would be traversing all attributes within the class, then filter values by class constraint, and then you can look up the list or dictionary to identify a column attribute's source. It's a bit cumbersome to prepare the whole column set for the inherited class(es) beforehand, however.
```python
list(
    cls_attr
    for cls_attr in (
        getattr(AttributeSetB, attr_name) for attr_name in dir(AttributeSetB)
    )
    if isinstance(cls_attr, Column)
)
```

**Comment by zzzeek** [comment, marked as answer]
OK so there's nothing "built in" for this, but you can identify which name came from which class by walking through `__mro__`.

the basic pattern that will do this is:

```py
attrname_to_cls = {
    attrname: cls
    for cls in MyModel.__mro__
    for attrname in dir(cls)
}
```


now there's some extra complications if you want to track attributes that are in more than one mixin/abstract, so below is a full demo that adds an extra step to try to figure that second part out.

```py
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import func
from sqlalchemy import inspect
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import AbstractConcreteBase
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declarative_mixin

Base = declarative_base()

@declarative_mixin
class TimestampMixin(object):
    created_at = Column(DateTime, default=func.now())

class AttributeSetA(Base):
    __abstract__ = True
    column_a_1 = Column(Integer)
    column_a_2 = Column(String)

    column_in_both_a_and_b = Column(String)

class AttributeSetB(AbstractConcreteBase, Base):
    __abstract__ = True
    column_b_1 = Column(Integer)
    column_b_2 = Column(String)

    column_in_both_a_and_b = Column(String)

class MyModel(AttributeSetA, AttributeSetB, TimestampMixin):
    __tablename__ = "my_model"
    id = Column(Integer, primary_key=True)

attrname_to_cls = {
    attrname: cls
    for cls in MyModel.__mro__
    for attrname in dir(cls)
}

# extra step to try to figure out attrs that are in two mixins
attrname_to_cls.update({
    attrname: cls
    for cls in reversed(MyModel.__mro__)
    for attrname in dir(cls)
    if not inspect(cls, raiseerr=False)
})

for attrname in inspect(MyModel).attrs.keys():
    print(f"Attribute {attrname} comes from class {attrname_to_cls[attrname]}")
```

**Reply by zzzeek** [reply]
note the tricky thing here is that the mapping process copied all the attributes to MyModel.  so if you see an attribute name in there, that's not necessarily where it came from.  hence the walking in forward/reverse order thing.

**Reply by pykenny** [reply]
Got it! Also thanks for providing the concrete example.

**Reply by MrYutz** [reply]
Thanks for this.....great example
